src/aerospike_clj/client.clj

```diff
@@ -121,29 +124,64 @@
+  (when (< MAX_BIN_NAME_LENGTH (.length bin-name))
+    (throw (Exception. (format "%s is %s characters. Bin names have to be <= 14 characters..." bin-name (.length bin-name)))))
+  (Bin. bin-name bin-value))
+
+(defn- ^Bin set-bin-as-null [^String bin-name]
+  (when (< MAX_BIN_NAME_LENGTH (.length bin-name))
+    (throw (Exception. (format "%s is %s characters. Bin names have to be <= 14 characters..." bin-name (.length bin-name)))))
+  (Bin/asNull bin-name))
+
 ;; get
 (defrecord AerospikeRecord [payload ^Integer gen ^Integer ttl])
 
 (defn- record->map [^Record record]
   (and record
-       (->AerospikeRecord
-         (get (.bins ^Record record) "")
-         ^Integer (.generation ^Record record)
-         ^Integer (.expiration ^Record record))))
+    (let [bins      (into {} (.bins ^Record record)) ;; converting from java.util.HashMap to a Clojure map
+          bin-names (mapv first bins)]
+      (->AerospikeRecord
+        (if (utils/single-bin? bin-names)
+          (utils/desanitize-bin-value (get bins ""))
+          (reduce-kv (fn [m k v]
+                       (assoc m (keyword k) (utils/desanitize-bin-value v)))
+            {}
+            bins))
+        ^Integer (.generation ^Record record)
+        ^Integer (.expiration ^Record record)))))
+
+(defn- map->multiple-bins [^IPersistentMap m]
+  (let [bins (for [[k v] m]
+               (create-bin (name k) (utils/sanitize-bin-value v)))]
+    (into-array Bin bins)))
+
+(defn- data->bins
+  "Function to identify whether `data` will be stored as a single or multiple bin record.
+  Only Clojure maps will default to multiple bins. Nested data structures are supported."
+  [data]
+  (if (map? data)
+    (map->multiple-bins data)
+    (into-array Bin [^Bin (Bin. "" (utils/sanitize-bin-value data))])))
 
 (defn get-single
   "Returns a single record: `(transcoder AerospikeRecord)`. The default transcoder is `identity`.
   Pass a `:policy` in `conf` to use a non-default `ReadPolicy`"
-  ([db index set-name] (get-single db index set-name {}))
-  ([db index set-name conf]
+  ([db index set-name] (get-single db index set-name [:all] {}))
+  ([db index set-name ^IPersistentVector bin-keys] (get-single db index set-name bin-keys {}))
+  ([db index set-name ^IPersistentVector bin-keys conf]
    (let [client (get-client db index)
          op-future (d/deferred)
-         start-time (System/nanoTime)]
+         start-time (System/nanoTime)
+         bin-names (mapv name bin-keys)] ;; bin-names can only be stored as strings in Aerospike
      (.get ^AerospikeClient client
            ^EventLoop (.next ^NioEventLoops (:el db))
            (reify-record-listener op-future)
            ^Policy (:policy conf)
-           (create-key (:dbns db) set-name index))
+           (create-key (:dbns db) set-name index)
+           ^"[Lcom.aerospike.client.Bin;" (if (= [:all] bin-keys)
+                                            (if (utils/single-bin? bin-names) bin-names nil)
```

Comment:
Yes. It seems that passing `nil` here returns all the `bins` in this record. I'm actually not a fan of the nested `if` statements since it makes this function a bit harder to reason about. 

I'll try to clean it up along when I am improving this function but suggestions are welcome

Reply:
sounds good. added comment and separated inner `if` statement